Rustfmt, clippy, fixes, and tests from trust-dns

little-dude opened the pull request:
I'm trying to figure out if is possible to use resolv-conf in trust dns (see [this issue](https://github.com/bluejekyll/trust-dns/issues/244)). The first step is to make sure all the trust-dns tests pass with resolv-conf so I started porting them. This is still work in progress but since there are already quite a lot of changes I thought I'd open a PR.

tailhook commented on src/lib.rs:
```diff
@@ -18,13 +22,106 @@ pub enum Network {
 ///
 /// Currently the options and defaults match those of linux/glibc
-#[derive(Clone, Debug)]
+#[derive(Default, Clone, Debug)]
```
Deriving default just to make the code below shorter is a bad idea. Especially because defaults for several parameters do not make sense. If we ever want to implement Default trait we would want to make it similar to current `Config::new`.

little-dude replied:
Ah I see. Makes sense. I'll implement `Default` manually then.

tailhook commented:
And if we're making a breaking change (I think I'm okay with that for now). We might also do two other things (not this PR, just thinking out loud):
1. Use `failure` crate instead of quick-error
2. Make `Config` extensible either by hiding all the fields under getter/setters or just by adding a private field (and waiting for `#[non_exhaustive]` to land).

What do you think?

little-dude commented:
Sorry this was not yet completely ready to merge.

- I removed the `Default` implementation
- I made the parser a bit more strict: it does not ignore directives or options it does not understand, and returns an error instead
- I derived `Eq` for `Config`. It's super convenient for tests and may be useful in general.

Note that some tests are still failing, but I'm not sure if it's because I expect the wrong thing or because the parsing is wrong. Basically, when we have:
```
sortlist 130.155.160.0/255.255.240.0 130.155.0.0
```

I think that the mask of the second ip should be `255.255.0.0`, but I'm not entirely sure. The parser currently says it's 255.255.255.255. The `man` is unclear about this:

> The netmask is optional and defaults to the natural netmask of the net.

little-dude commented on tests/lib.rs:
```diff
@@ -0,0 +1,197 @@
+            Ipv4Addr::new(255, 255, 240, 0),
+        ),
+        // This fails currently
+        Network::V4(Ipv4Addr::new(130, 155, 0, 0), Ipv4Addr::new(255, 255, 0, 0)),
```
This is the failing test. Will look into it tonight.

little-dude replied:
If ["DNS and Bind"](https://books.google.com/books?id=HggtWI1ShvMC&pg=PT133&lpg=PT133&dq=sortlist+mask+resolv&source=bl&ots=PTKqTUTz0v&sig=KpUKVZDtUyK3i2nsIS0Qf69TedI&hl=en&sa=X&ved=0ahUKEwiBmd-eouXXAhVK0oMKHQ8BB1gQ6AEINTAC#v=onepage&q=sortlist%20mask%20resolv&f=false) is correct, the expected mask is `255.255.0.0`, so this is a bug.

> The argument after the slash is the subnet mask for the subnet in question. To prefer an entire network, you can omit the slash and the subnet mask
>  `sortlist 128.32.0.0`
> The resolver then assumes you mean the entire network 128.32/16. (The resolver derives the default unsubnetted net mask for the network from the first two bits of the IP address.)

I assume in the last sentence they meant "bytes" not "bits", it does not make sense otherwise.

little-dude commented:
> Make Config extensible either by hiding all the fields under getter/setters or just by adding a private field (and waiting for #[non_exhaustive] to land).

I think there is an issue currently with `Config`. After we parse a file, there is no way to make a difference between an attribute that was set to its default value and an attribute that is set in the file but to the same value than the default value. For instance, if I have `config.timeout == 5 ` is it because it was _not_ specified in my file, or because it _was_ specified, and explicitely set to 5?

I can imagine that the ambiguity will be a problem for resolvers that want to use different default values.

little-dude commented:
Regarding `failure` I don't really have a technical opinion because I never used it yet. But it has been presented to the community like the next step after `quick_error` and `error_chain` so it probably makes sense to use it.

little-dude commented:
All right I think I'm done with this PR. I fixed the failing test in 004609a, most of the relevant tests have been ported, and I added a few new ones. The last commit creates a `config` and `ip` modules. If you think it's too many changes for this PR, I'll back it out, since it's not strictly necessary.

little-dude commented on src/grammar.rs:
```diff
@@ -34,22 +34,58 @@ quick_error!{
-
 fn ip_v4_netw(val: &str) -> Result<Network, AddrParseError> {
     let mut pair = val.splitn(2, '/');
-    let ip = pair.next().unwrap().parse()?;
```
These changes make the function a bit ugly. Maybe we could refactor it later.

bluejekyll commented:
> The ambiguity might be a problem for resolvers that want to use different default values. @bluejekyll would that be a problem for trust-dns for instance?

The only issue I see on the Resolver side is if this comes into conflict with CLI parameters, vs. resolv.conf, or some other config file. I don't foresee an issue there, but that's always something we can fixup later as necessary.

bluejekyll commented:
> Regarding failure I don't really have a technical opinion because I never used it yet. But it has been presented to the community like the next step after quick_error and error_chain so it probably makes sense to use it.

I've only briefly looked at `failure`. I currently have no opinion. If it becomes the general trend, I'd be a big supporter. Does `failure` require nightly or is it ok on stable? Personally I'd suggest holding off on a move to `failure` until the direction there is a little more clear.

tailhook commented:
Merged in. Thanks!

> Does failure require nightly or is it ok on stable?

Works on stable.

> Personally I'd suggest holding off on a move to failure until the direction there is a little more clear.

Okay.

tailhook commented:
Added a short list of things to discuss/implement for 0.5.0 in #3.
